crates/orchestrator/src/tests/jobs/mod.rs

```diff
@@ -20,3 +20,194 @@ async fn create_job_fails_job_already_exists() {
 async fn create_job_fails_works_new_job() {
     // TODO
 }
+
+#[cfg(test)]
+mod job_handler_tests {
+    use std::collections::HashMap;
+    use std::sync::Arc;
+
+    use mockall::predicate::eq;
+    use rstest::rstest;
+    use uuid::Uuid;
+
+    use crate::config::config;
+    use crate::jobs::constants::{JOB_PROCESS_ATTEMPT_METADATA_KEY, JOB_VERIFICATION_ATTEMPT_METADATA_KEY};
+    use crate::jobs::job_handler_factory::mock_factory;
+    use crate::jobs::types::{ExternalId, JobItem, JobStatus, JobType, JobVerificationStatus};
+    use crate::jobs::{create_job, process_job, verify_job, Job, MockJob};
+    use crate::tests::common::drop_database;
+    use crate::tests::config::TestConfigBuilder;
+
+    #[rstest]
+    #[case(JobType::SnosRun, false, true)]
+    #[case(JobType::ProofCreation, true, true)]
+    #[should_panic]
+    #[case(JobType::ProofRegistration, false, false)]
+    #[tokio::test]
+    async fn test_create_job_handler(
+        #[case] job_type: JobType,
+        #[case] job_exists_in_db: bool,
+        #[case] job_implemented: bool,
+    ) {
+        let job_item = JobItem {
+            id: Uuid::new_v4(),
+            internal_id: "0".to_string(),
+            job_type: job_type.clone(),
+            status: JobStatus::Created,
+            external_id: ExternalId::Number(0),
+            metadata: Default::default(),
+            version: 0,
+        };
+
+        let mut job_handler = MockJob::new();
+        if job_implemented && !job_exists_in_db {
+            // Expecting for create_job handler for that particular job.
+            let job_item_clone = job_item.clone();
+            job_handler.expect_create_job().times(1).returning(move |_, _, _| Ok(job_item_clone.clone()));
+        }
+
+        TestConfigBuilder::new().build().await;
+        drop_database().await.unwrap();
+
+        let config = config().await;
+        let database_client = config.database();
+
+        if job_exists_in_db {
+            database_client.create_job(job_item).await.unwrap();
+        }
+
+        if job_implemented && !job_exists_in_db {
```

Comment:
i think let's just create different tests instead of if else based on cases. currently, it's hard to read what's being tested

Reply:
applies to all tests below as well

Reply:
we can do one PR for every function - all tests of `process_job`, all tests of `verify_job` etc.